src/main/java/iterator/RichIterator.java

```diff
@@ -10,37 +13,57 @@
      * @return true if there are no more elements, false otherwise
      */
     default boolean isEmpty() {
-        throw new NotImplementedException();
+        return !this.hasNext();
     }
 
     /**
      * @return number of elements
      */
     default int length() {
-        throw new NotImplementedException();
+        int counter = 0;
+        while (this.hasNext()) {
+            this.next();
+            counter++;
+        }
+        return counter;
     }
 
     /**
      * @return the last element of the iterator
      * @throws NoSuchElementException if the iterator is empty
      */
     default A last() {
-        throw new NotImplementedException();
+        A last = this.next();
+        while (this.hasNext()) {
+            last = this.next();
+        }
+        return last;
     }
 
     /**
      * @return the last element if not empty.
      */
     default Optional<A> lastOptional() {
-        throw new NotImplementedException();
+        Optional<A> last = Optional.empty();
+        while (this.hasNext()) {
+            last = Optional.ofNullable(this.next());
+        }
+        return last;
```

Comment:
duplicate

Reply:
I don't understand what do you mean...
What duplicate?

Reply:
@shiloHac 
this and the function above are the same give or take